flink-connector-dynamodb/src/main/java/org/apache/flink/streaming/connectors/dynamodb/sink/DynamoDbSinkWriter.java

```diff
@@ -0,0 +1,216 @@
+ * internals of this sink writer may be found in {@link AsyncSinkWriter}.
+ *
+ * <p>The {@link DynamoDbAsyncClient} used here may be configured in the standard way for the AWS
+ * SDK 2.x. e.g. the provision of {@code AWS_REGION}, {@code AWS_ACCESS_KEY_ID} and {@code
+ * AWS_SECRET_ACCESS_KEY} through environment variables etc.
+ */
+@Internal
+class DynamoDbSinkWriter<InputT> extends AsyncSinkWriter<InputT, DynamoDbWriteRequest> {
+    private static final Logger LOG = LoggerFactory.getLogger(DynamoDbSinkWriter.class);
+
+    /* A counter for the total number of records that have encountered an error during put */
+    private final Counter numRecordsSendErrorsCounter;
+
+    /* The sink writer metric group */
+    private final SinkWriterMetricGroup metrics;
+
+    private final DynamoDbAsyncClient client;
+    private final boolean failOnError;
+    private final String tableName;
+
+    private List<String> overwriteByPKeys;
+
+    public DynamoDbSinkWriter(
+            ElementConverter<InputT, DynamoDbWriteRequest> elementConverter,
+            InitContext context,
+            int maxBatchSize,
+            int maxInFlightRequests,
+            int maxBufferedRequests,
+            long maxBatchSizeInBytes,
+            long maxTimeInBufferMS,
+            long maxRecordSizeInBytes,
+            boolean failOnError,
+            String tableName,
+            List<String> overwriteByPKeys,
+            Properties dynamoDbClientProperties,
+            Collection<BufferedRequestState<DynamoDbWriteRequest>> states) {
+        super(
+                elementConverter,
+                context,
+                maxBatchSize,
+                maxInFlightRequests,
+                maxBufferedRequests,
+                maxBatchSizeInBytes,
+                maxTimeInBufferMS,
+                maxRecordSizeInBytes,
+                states);
+        this.failOnError = failOnError;
+        this.tableName = tableName;
+        this.overwriteByPKeys = overwriteByPKeys;
+        this.metrics = context.metricGroup();
+        this.numRecordsSendErrorsCounter = metrics.getNumRecordsSendErrorsCounter();
+        this.client = AWSDynamoDbUtil.createClient(dynamoDbClientProperties);
+    }
+
+    @Override
+    protected void submitRequestEntries(
+            List<DynamoDbWriteRequest> requestEntries,
+            Consumer<List<DynamoDbWriteRequest>> requestResultConsumer) {
+
+        TableRequestsContainer container = new TableRequestsContainer(overwriteByPKeys);
+        for (DynamoDbWriteRequest requestEntry : requestEntries) {
+            container.put(convertToWriteRequest(requestEntry));
+        }
+
+        CompletableFuture<BatchWriteItemResponse> future =
+                client.batchWriteItem(
+                        BatchWriteItemRequest.builder()
+                                .requestItems(
+                                        ImmutableMap.of(tableName, container.getRequestItems()))
+                                .build());
```

Comment:
As an optimization, we could completely skip `PrimaryKeyBuilder.build` when  `CollectionUtil.isNullOrEmpty(overwriteByPKeys)`. This would remove the UUID generation and map overhead. Something like this:

```
        final List<WriteRequest> writeRequests;
        
        if (CollectionUtil.isNullOrEmpty(overwriteByPKeys)) {
            writeRequests = new ArrayList<>();
            for (DynamoDbWriteRequest requestEntry : requestEntries) {
                writeRequests.put(convertToWriteRequest(requestEntry));
            }
        } else {
            TableRequestsContainer container = new TableRequestsContainer(overwriteByPKeys);
            for (DynamoDbWriteRequest requestEntry : requestEntries) {
                container.put(convertToWriteRequest(requestEntry));
            }

            writeRequests = container.getRequestItems();
        }

        CompletableFuture<BatchWriteItemResponse> future =
                client.batchWriteItem(
                        BatchWriteItemRequest.builder()
                                .requestItems(ImmutableMap.of(tableName, writeRequests))
                                .build());

```